web/src/hotkey.js

```diff
@@ -121,51 +121,144 @@ const keydown_either_mappings = {
+    s: {name: "narrow_by_recipient", message_view_only: true}, // 's'
+    t: {name: "open_recent_topics", message_view_only: true}, // 't'
+    u: {name: "show_sender_info", message_view_only: true}, // 'u'
+    v: {name: "show_lightbox", message_view_only: true}, // 'v'
+    w: {name: "query_users", message_view_only: true}, // 'w'
+    x: {name: "compose_private_message", message_view_only: true}, // 'x'
+    z: {name: "zoom_to_message_near", message_view_only: true}, // 'z'
 };
 
+const shifted_codes_to_keys = {
+    Backquote: "~",
+    Backslash: "|",
+    BracketLeft: "{",
+    BracketRight: "}",
+    Comma: "<",
+    Digit0: ")",
+    Digit1: "!",
+    Digit2: "@",
+    Digit3: "#",
+    Digit4: "$",
+    Digit5: "%",
+    Digit6: "^",
+    Digit7: "&",
+    Digit8: "*",
+    Digit9: "(",
+    Equal: "+",
+    Minus: "_",
+    Period: ">",
+    Quote: '"',
+    Semicolon: ":",
+    Slash: "?",
+};
+
+const unshifted_codes_to_keys = {
+    Backquote: "`",
+    Backslash: "\\",
+    BracketLeft: "[",
+    BracketRight: "]",
+    Comma: ",",
+    Digit0: "0",
+    Digit1: "1",
+    Digit2: "2",
+    Digit3: "3",
+    Digit4: "4",
+    Digit5: "5",
+    Digit6: "6",
+    Digit7: "7",
+    Digit8: "8",
+    Digit9: "9",
+    Equal: "=",
+    Minus: "-",
+    Period: ".",
+    Quote: "'",
+    Semicolon: ";",
+    Slash: "/",
+};
+
+function get_key_from_event(e) {
+    const {key, code, shiftKey} = e;
+
+    // If the layout is QWERTY, DVORAK, AZERTY etc (any latin layout),
+    // use e.which, since the inputted character likely has a valid ASCII,
+    // which is what we use to match our shortcuts.
+    if (key.codePointAt(0) < 256) {
+        return key;
+    }
+
+    // Our layout is not latin. Example: Cyrillic is in the early 1000s range.
+
+    // There is no easy way to figure out when capslock is on, so we'll pretend
+    // it is always off.
+
+    // If it is an alphabet key, it'll be of the format: "KeyA" for both A and a.
+    // Thus, we assume capslock is off for our uppercase/lowercase behavior.
+    if (code.startsWith("Key") && code.length === 4) {
+        let inferred_key = code.charAt(3);
+        if (!shiftKey) {
+            inferred_key = inferred_key.toLowerCase();
+        }
+        return inferred_key;
+    }
+
+    // We have a map of code to key pairs, which we'll use here.
+
+    // NOTE: Since this code is checked at last, for Cyrillic keyboards and the like
+    // we will interpret the keys as they are in that layout.
+
+    // For example, pressing shift+/ on a Cyrillic layout would input , instead of ?.
+    if (shiftKey) {
+        if (shifted_codes_to_keys[code]) {
+            return shifted_codes_to_keys[code];
+        }
+    } else {
+        if (unshifted_codes_to_keys[code]) {
+            return unshifted_codes_to_keys[code];
+        }
+    }
+
+    // If our logic above didn't capture it, let's just bail and return the original.
+    return key;
```

Comment:
What are some situations where we fall through to this point?

Reply:
Probably the Euro symbol? Or some other currency symbol?

It'd fall through to here if it wouldn't be a character key (KeyA, KeyB, etc), wouldn't be covered in the original ASCII list (newer/rarer currency symbols fit this) and aren't in our hardcoded keys.